src/flow-worker.c

```diff
@@ -170,7 +168,16 @@ TmEcode FlowWorker(ThreadVars *tv, Packet *p, void *data, PacketQueue *preq, Pac
         FlowHandlePacket(tv, fw->dtv, p);
         if (likely(p->flow != NULL)) {
             DEBUG_ASSERT_FLOW_LOCKED(p->flow);
-            FlowUpdate(p);
+            int state = SC_ATOMIC_GET(p->flow->flow_state);
+            FlowUpdate(p, state);
+            switch (state) {
```

Comment:
move this switch into FlowUpdate and make it return a value that we can check for discarding?